core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala

```diff
@@ -129,11 +135,22 @@ private[sql] object PartitionedPhysicalScan {
         val (a, f) = scanBuilderArgs
         val baseTableRDD = table.buildRowBufferRDD(Array.empty,
           a.map(_.name).toArray, f.toArray, useResultSet = false)
+
+        def resolveCol(left: Attribute, right: AttributeReference) =
+          columnScan.sqlContext.sessionState.analyzer.resolver(left.name, right.name)
+
         val rowBufferScan = RowTableScan(output, baseTableRDD, numBuckets,
           Seq.empty, table)
-        val bufferExchange = ShuffleExchange(columnScan.outputPartitioning,
-          rowBufferScan)
-        ZipPartitionScan(columnScan, bufferExchange)
+        val otherPartKeys = partitionColumns.map(_.transform {
+          case a: AttributeReference => rowBufferScan.output.find(resolveCol(_, a)).getOrElse {
+            throw new AnalysisException(s"RowBuffer output column $a not found in " +
+                s"${rowBufferScan.output.mkString(",")}")
+          }
+        })
+        assert(columnScan.outputPartitioning.satisfies(
+          ClusteredDistribution(columnScan.partitionColumns)))
+        ZipPartitionScan(columnScan, columnScan.partitionColumns,
+          rowBufferScan, otherPartKeys)
```

Comment:
I think I understood your question... generated RDD as in the second RDD right ?
I haven't followed the partitioning till preferredLocation of generatedRDD but I assumed if number of partitions remain as columnScan, we should be good around referredLocation. I have verified that only. InputRDDs second one adapts to the first one because of second one being rdd of ShuffleExchange. Right now RowTableScan RDDs don't have preferred location or partition columns mentioned that will fetch data from all over & that is the intent no ?